### ISSUE TITLE
Feature Request: Always remember passphrase

### ISSUE THREAD

**Issue by jgornick** [tejado/obsidian-gpgCrypt#10]
First off, great plugin! It's exactly what I was looking for coming from Joplin and the encryption done there.

I wanted to see if it was possible to always remember the passphrase for the key? I see you can provide `0` to disable, but I'm assuming that's to disable remembering the passphrase as I've been prompted to always provide a password.

Any help would be greatly appreciated! Thank you!

**Comment by tejado** [comment]
Thanks for the feedback. 
You could set a time of maximum one month. Would this be enough?
Please keep in mind that its only remembering the passphrase when its running. Closing Obisdian will not write the passphrase to disk.

**Comment by rubaboo** [comment]
UPDATE: Upon finding a clue in discussion #7, my guess is the relevant settings in gpg4win are these:

<img src="https://github.com/tejado/obsidian-gpgCrypt/assets/212368/e04e4eb5-ed2f-4225-b1a0-3d3ff3523a4b" width="623">

Happy to report that after closing Obsidian and waiting for a few minutes, I got prompted for the passphrase again when opening the vault.

@tejado I think it would be beneficial to have an blurb in Readme for people not familiar with gpg and it's passphrase caching feature. Also, "Set to 0 to disable" should instead say something like "Set to 0 to remember until Obsidian is closed (see "About passphrase caching" caveat)".

Finally, I'd like to say this is really a game-changing plugin. I'm surprised none of the YouTube influencers featured it yet. 😃
Hope Obsidian does not break it with future updates.

**Comment by rubaboo** [comment]
I understand what you are saying, however, I still think it makes sense to be able to let Obsidian remember it's passphrase for the entire time it is running, or for a longer duration anyways, without leaving whatever else one might be using GPG for more exposed.

**Comment by Aetherinox** [comment]
> without leaving whatever else one might be using GPG for more exposed.

Not sure I get what you mean. Excluding any GUI based applications that can help you edit GPG settings using an interface because of the default config file, this plugin much like all other GPG-based technologies require that GPG be installed.

GPG remembering your passphrase is handled by the agent itself, which has to run as a service even if you want GPG in obsidian to work.

Changing this involves editing `~/.gnupg/gpg-agent.conf` and simply modifying the default value

``
# in seconds
default-cache-ttl 3600
``

This setting is global across any service that needs GPG, such as this Obsidian plugin, SSH, Github signing, Yubikey GPG signing / encrypting, etc.

Once the cache time on GPG itself runs out, it takes priority over any other application and it will prompt you for your passphrase again in order for the agent to communicate with the program.

And if there were to be some type of passphrase remembered by this obsidian plugin, it would have to be cached somewhere, potentially in raw format, which is a huge security risk. And then plugin would have to have some way to inject the passphrase into the GPG agent when it requests for the user to input that passphrase again.

With the config above, that setting is only used by the gpg agent, which has to run for the Obsidian plugin to even operate, so no additional programs need to run for that setting to take affect. If you kill the gpg service, then the Obsidian plugin can't communicate with GPG at all.

**Comment by rubaboo** [comment]
Hi, thanks for the explanation. I totally thought it was the plugin remembering the passphrase. If it does not, then I agree, the setting is superfluous. I'm not proficient enough to comb through the code to find out.

Still hope it does remember the passphrase. The size of the security risk depends on my threat model. If I only want to prevent someone snooping in my vault files at rest, and I don't use that particular key for anything else, then the risk is quite acceptable.

**Comment by Aetherinox** [comment]
Sure thing.

Yeah, the handling of the passphrase is completely up to the GPG agent running on your system. I'm not even sure how the developer could do a "remember passphrase" feature for Obsidian, without the risk of the passphrase being retrievable.

Obviously GPG command-line has support for specifying the GPG key passphrase when executing sign / generate / verify commands, but having the plain-text passphrase just sitting there would be a huge issue in my eyes. 

To increase the amount of time GPG members your passphrase, you can edit the settings I listed above. Such as for mine, I have the following:

Windows: `C:\Users\USERNAME\AppData\Roaming\gnupg\gpg-agent.conf`
Linux: `/home/USERNAME/.gnupg/gpg-agent.conf`

```
enable-putty-support
enable-ssh-support
use-standard-socket
default-cache-ttl-ssh 60
max-cache-ttl-ssh 120
default-cache-ttl 28800 # gpg key cache time
max-cache-ttl 28800 # max gpg key cache time
pinentry-program "C:\Program Files (x86)\Gpg4win\bin\pinentry.exe"
allow-loopback-pinentry
allow-preset-passphrase
pinentry-timeout 0
```

For SSH, I have the passphrase expire after a few minutes. However, for stuff like encrypting / signing, which is what this Obsidian plugin does, I have the cache time set to 8 hours / 28800 seconds.

You can remove the `pinentry` settings if you don't need them. There's an explanation of pinetry at https://velvetcache.org/2023/03/26/a-peek-inside-pinentry/ or https://www.gnu.org/software/emacs/manual/html_node/epa/GnuPG-Pinentry.html

After you edit the config, simply restart GPG by opening terminal or windows command prompt and executing

```shell
gpgconf --kill gpg-agent
```

**Comment by tejado** [comment]
Hi all,
so the current minimum "remember passphrase" time is 10 seconds due to changes in the underlying Obsidian read methods. 

Instead of having a very high value for "remember passphrase", would you still like to have a separate setting to cache the passphrase infinitely in memory?

**Comment by rubaboo** [comment]
> You could set a time of maximum one month. Would this be enough?

If this is still the case, then it's fine. I joined this conversation only because I got very confused by GPG caching the passphrase, so I wanted to clarify what exactly `Remember passphrase` `0` means. I'm good now.

**Comment by tejado** [comment]
> > You could set a time of maximum one month. Would this be enough?
> 
> If this is still the case, then it's fine. I joined this conversation only because I got very confused by GPG caching the passphrase, so I wanted to clarify what exactly `Remember passphrase` `0` means. I'm good now.

Yeah, this was probably my fault as I did not hide the passphrase cache when GPG CLI backend is selected. This got now corrected.
For GPG CLI, the limits are configured externally in the gpg config outside of Obsidian and my gpgCrypt plugin.

**Comment by tejado** [comment]
Closing as this is possible by:
- Setting a high value for "Remember passprhase" (e.g. ~11.5 days is a value of 999999 seconds)
or
- Using gpg CLI wrapper with a specific passphrase cache configuration.
